File: test/fanout.test.ts
import { describe, expect, it, vi, beforeEach } from 'vitest';
import type { WorkerResult } from '../src/types.js';
import { RunLogger } from '../src/logger.js';
import { mkdtempSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

vi.mock('../src/workers/index.js', () => ({ getWorker: vi.fn() }));
vi.mock('../src/git/worktree.js', () => ({ createWorktree: vi.fn() }));

import { getWorker } from '../src/workers/index.js';
import { createWorktree } from '../src/git/worktree.js';
import { runFanout } from '../src/workers/fanout.js';
import { planFromTicket } from '../src/planner.js';

const mockGetWorker = vi.mocked(getWorker);
const mockCreate = vi.mocked(createWorktree);

const plan = planFromTicket({
  id: 'ENG-9',
  title: 'Add GET /health endpoint returning status JSON',
  description: 'Endpoint returns service status as JSON including uptime.',
  labels: [],
  acceptanceCriteria: ['returns 200'],
});

const result = (over: Partial<WorkerResult> = {}): WorkerResult => ({
  exitCode: 0,
  events: [],
  resultText: null,
  sessionId: null,
  durationMs: 1000,
  timedOut: false,
  ...over,
});

function workerMock(exitCode: number) {
  return { name: 'claude-code' as const, spawn: vi.fn().mockResolvedValue(result({ exitCode })) };
}

describe('runFanout', () => {
  beforeEach(() => {
    mockGetWorker.mockReset();
    mockCreate.mockReset();
  });

  const logDir = () => mkdtempSync(join(tmpdir(), 'da-fan-'));
  const logger = (dir: string) => new RunLogger(dir);

  it('prefers the leg whose tests pass', async () => {
    const dir = logDir();
    try {
      const claude = workerMock(0);
      const open = workerMock(0);
      mockGetWorker.mockImplementation((n) => (n === 'claude-code' ? claude : open) as never);
      mockCreate.mockImplementation(async (_r, id) => ({
        worktreePath: `/tmp/leg-${id}`,
        branch: `devagent/${id}`,
      }));
      const score = vi.fn().mockResolvedValue(false).mockResolvedValueOnce(true); // claude leg passes

      const winner = await runFanout(plan, ['claude-code', 'opencode'], logger(dir), {
        repoPath: '/tmp/repo',
        timeoutMs: 5000,
        scoreLeg: score,
      });
      expect(winner?.worker).toBe('claude-code');
      expect(score).toHaveBeenCalledTimes(3); // opencode leg gets one flaky rerun
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });

  it('returns null when every leg fails', async () => {
    const dir = logDir();
    try {
      mockGetWorker.mockReturnValue(workerMock(1) as never);
      mockCreate.mockRejectedValue(new Error('branch exists')); // forces cwd fallback, still exit 1
      const winner = await runFanout(plan, ['claude-code'], logger(dir), { repoPath: '/tmp/r', timeoutMs: 5000 });
      expect(winner).toBeNull();
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });

  it('falls back to any usable leg when scoring is unavailable', async () => {
    const dir = logDir();
    try {
      mockGetWorker.mockImplementation((n) =>
        n === 'claude-code' ? workerMock(1) : workerMock(0),
      ) as never;
      mockCreate.mockRejectedValue(new Error('no worktrees here'));
      const winner = await runFanout(plan, ['claude-code', 'opencode'], logger(dir), {
        repoPath: '/tmp/r',
        timeoutMs: 5000,
      });
      expect(winner?.worker).toBe('opencode');
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });

  it('rescues a leg whose tests fail once then pass on rerun, marked flaky', async () => {
    const dir = logDir();
    try {
      mockGetWorker.mockReturnValue(workerMock(0) as never);
      mockCreate.mockImplementation(async (_r, id) => ({
        worktreePath: `/tmp/leg-${id}`,
        branch: `devagent/${id}`,
      }));
      const score = vi.fn().mockResolvedValue(true).mockResolvedValueOnce(false);

      const winner = await runFanout(plan, ['opencode'], logger(dir), {
        repoPath: '/tmp/repo',
        timeoutMs: 5000,
        scoreLeg: score,
      });
      expect(winner?.worker).toBe('opencode');
      expect(winner?.flaky).toBe(true);
      expect(score).toHaveBeenCalledTimes(2);
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });

  it('prefers a clean pass over a flaky pass', async () => {
    const dir = logDir();
    try {
      const claude = workerMock(0);
      const open = workerMock(0);
      mockGetWorker.mockImplementation((n) => (n === 'claude-code' ? claude : open) as never);
      mockCreate.mockImplementation(async (_r, id) => ({
        worktreePath: `/tmp/leg-${id}`,
        branch: `devagent/${id}`,
      }));
      // Route by worktree path: claude leg passes cleanly, opencode leg fails
      // on its first run and passes on the rerun.
      const calls: Record<string, number> = {};
      const routed = async (wt: string): Promise<boolean> => {
        calls[wt] = (calls[wt] ?? 0) + 1;
        if (wt.includes('claude-code')) return true;
        return calls[wt] >= 2; // fails on first run, passes on rerun
      };

      const winner = await runFanout(plan, ['claude-code', 'opencode'], logger(dir), {
        repoPath: '/tmp/repo',
        timeoutMs: 5000,
        scoreLeg: routed,
      });
      expect(winner?.worker).toBe('claude-code');
      expect(winner?.flaky ?? false).toBe(false);
      expect(calls['/tmp/leg-ENG-9-opencode']).toBe(2);
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });

  it('keeps a persistently failing leg failed after the single rerun', async () => {
    const dir = logDir();
    try {
      mockGetWorker.mockReturnValue(workerMock(0) as never);
      mockCreate.mockImplementation(async (_r, id) => ({
        worktreePath: `/tmp/leg-${id}`,
        branch: `devagent/${id}`,
      }));
      const score = vi.fn().mockResolvedValue(false);

      const winner = await runFanout(plan, ['opencode'], logger(dir), {
        repoPath: '/tmp/repo',
        timeoutMs: 5000,
        scoreLeg: score,
      });
      // Sole usable leg still wins (existing semantics), but stays unflagged
      // and ranked as failed — the guard only rescues pass-after-rerun legs.
      expect(winner?.worker).toBe('opencode');
      expect(winner?.testsPassed).toBe(false);
      expect(winner?.flaky ?? false).toBe(false);
      expect(score).toHaveBeenCalledTimes(2);
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });
});

File: internal/workers/fanout_test.go
// Go port of test/fanout.test.ts: winner ranking policy — one flaky
// rerun, clean pass outranks a flaky rescue, null when every leg fails.

package workers

import "testing"

func boolPtr(b bool) *bool { return &b }

// fakeWorkerResult builds a success-shaped WorkerResult with the given
// exit code (the TS workerMock(exitCode)).
func fakeWorkerResult(exitCode int) WorkerResult {
	return WorkerResult{ExitCode: exitCode, DurationMs: 1000}
}

// runFanoutLegs mirrors the per-leg portion of the TS runFanout: spawn
// (via injected runner), score with the flaky guard, derive the failure
// class, then rank across usable legs. Worktree creation failure falls
// back to the repo path (the TS catch branch).
type fanoutLegFixture struct {
	worker   string
	exitCode int
	scores   []*bool // per scoreLeg call for this leg
	timeouts []bool
}

func runFanoutLegs(fixtures []fanoutLegFixture) ([]FanoutLeg, []int) {
	legs := make([]FanoutLeg, 0, len(fixtures))
	callCounts := make([]int, len(fixtures))
	for i, f := range fixtures {
		worktree := "/tmp/leg-" + f.worker
		result := fakeWorkerResult(f.exitCode)
		ok := !result.TimedOut && result.ExitCode == 0
		timedOut := len(f.timeouts) > 0 && f.timeouts[0]
		var score FanoutScoreLeg
		if f.scores != nil {
			score = func(string, int) *bool {
				callCounts[i]++
				return f.scores[callCounts[i]-1]
			}
		}
		testsPassed, flaky := ScoreFanoutLeg(ok, worktree, score)
		legs = append(legs, FanoutLeg{
			Worker:       f.worker,
			WorktreePath: worktree,
			Ok:           ok,
			TestsPassed:  testsPassed,
			Flaky:        flaky,
			DurationMs:   1000,
			FailureClass: fanoutLegFailureClass(ok, timedOut, testsPassed),
		})
	}
	return legs, callCounts
}

func TestFanout_PrefersTheLegWhoseTestsPass(t *testing.T) {
	// claude leg passes; opencode leg fails once then passes on the rerun.
	legs, counts := runFanoutLegs([]fanoutLegFixture{
		{worker: "claude-code", exitCode: 0, scores: []*bool{boolPtr(true)}},
		{worker: "opencode", exitCode: 0, scores: []*bool{boolPtr(false), boolPtr(true)}},
	})
	usable := FilterUsableLegs(legs)
	if len(usable) != 2 {
		t.Fatalf("usable = %+v", usable)
	}
	winner := SelectFanoutWinner(usable)
	if winner.Worker != "claude-code" {
		t.Fatalf("winner = %+v", winner)
	}
	// opencode leg gets one flaky rerun (2 calls) + claude's 1.
	if counts[1] != 2 {
		t.Fatalf("opencode score calls = %d, want 2", counts[1])
	}
}

func TestFanout_ReturnsNilWhenEveryLegFails(t *testing.T) {
	// Worktree failure forces the cwd fallback; exit 1 keeps ok=false.
	legs, _ := runFanoutLegs([]fanoutLegFixture{
		{worker: "claude-code", exitCode: 1},
	})
	if usable := FilterUsableLegs(legs); len(usable) != 0 {
		t.Fatalf("expected no usable legs, got %+v", usable)
	}
	// The TS runFanout returns null — no winner selectable.
	if legs[0].FailureClass != "worker-error" {
		t.Fatalf("failureClass = %q, want worker-error", legs[0].FailureClass)
	}
}

func TestFanout_FallsBackToAnyUsableLegWhenScoringUnavailable(t *testing.T) {
	// No scoreLeg: claude fails (exit 1), opencode passes.
	legs, counts := runFanoutLegs([]fanoutLegFixture{
		{worker: "claude-code", exitCode: 1},
		{worker: "opencode", exitCode: 0},
	})
	winner := SelectFanoutWinner(FilterUsableLegs(legs))
	if winner.Worker != "opencode" {
		t.Fatalf("winner = %+v", winner)
	}
	if counts[1] != 0 {
		t.Fatalf("unscored leg must not call scoreLeg: %d", counts[1])
	}
	if legs[0].TestsPassed != nil {
		t.Fatalf("unscored leg testsPassed must be nil, got %v", *legs[0].TestsPassed)
	}
}

func TestFanout_RescuesFlakyLeg(t *testing.T) {
	// Tests fail once then pass on rerun — marked flaky, still wins.
	legs, counts := runFanoutLegs([]fanoutLegFixture{
		{worker: "opencode", exitCode: 0, scores: []*bool{boolPtr(false), boolPtr(true)}},
	})
	winner := SelectFanoutWinner(FilterUsableLegs(legs))
	if winner.Worker != "opencode" {
		t.Fatalf("winner = %+v", winner)
	}
	if !winner.Flaky {
		t.Fatalf("winner must be flagged flaky: %+v", winner)
	}
	if counts[0] != 2 {
		t.Fatalf("score calls = %d, want 2", counts[0])
	}
}

func TestFanout_PrefersCleanPassOverFlakyPass(t *testing.T) {
	// Claude passes cleanly; opencode fails then passes on the rerun.
	// The clean pass must outrank the flaky rescue.
	legs, counts := runFanoutLegs([]fanoutLegFixture{
		{worker: "claude-code", exitCode: 0, scores: []*bool{boolPtr(true)}},
		{worker: "opencode", exitCode: 0, scores: []*bool{boolPtr(false), boolPtr(true)}},
	})
	winner := SelectFanoutWinner(FilterUsableLegs(legs))
	if winner.Worker != "claude-code" {
		t.Fatalf("clean pass must outrank flaky: %+v", winner)
	}
	if winner.Flaky {
		t.Fatalf("winner must not be flaky: %+v", winner)
	}
	if counts[1] != 2 {
		t.Fatalf("opencode score calls = %d, want 2", counts[1])
	}
}

func TestFanout_KeepsPersistentlyFailingLegFailed(t *testing.T) {
	// A sole usable leg still wins (existing semantics) but stays unflagged
	// and ranked as failed — the guard only rescues pass-after-rerun legs.
	legs, counts := runFanoutLegs([]fanoutLegFixture{
		{worker: "opencode", exitCode: 0, scores: []*bool{boolPtr(false), boolPtr(false)}},
	})
	winner := SelectFanoutWinner(FilterUsableLegs(legs))
	if winner.Worker != "opencode" {
		t.Fatalf("sole usable leg still wins: %+v", winner)
	}
	if winner.TestsPassed == nil || *winner.TestsPassed {
		t.Fatalf("testsPassed must stay false: %+v", winner)
	}
	if winner.Flaky {
		t.Fatalf("flaky must stay false: %+v", winner)
	}
	if counts[0] != 2 {
		t.Fatalf("exactly one rerun allowed: %d calls", counts[0])
	}
}

func TestFanoutWorkerSuffix(t *testing.T) {
	if got := WorkerSuffix("claude-code", 0); got != "claude-code" {
		t.Fatalf("suffix = %q", got)
	}
	// non-alnum characters stripped (TS replace semantics)
	if got := WorkerSuffix("we!rd@name", 3); got != "werdname" {
		t.Fatalf("suffix = %q, want werdname", got)
	}
	if got := WorkerSuffix("!!!", 7); got != "leg7" {
		t.Fatalf("empty suffix must fall back to leg7, got %q", got)
	}
}
